### PR TITLE
Town window implementation

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Not much to see here yet, just getting things ready.

### CODE REVIEW

File: src/openloco/windows/townwnd.cpp [hunk 1]
```diff
@@ -231,13 +234,52 @@ namespace openloco::ui::windows::town
...
+            self->var_848 = edx;
+            self->var_84C = ecx;
+
+            // 0x00499B39 start
+            if (self->viewports[0] == nullptr)
+            {
+                auto widget = &self->widgets[widx::viewport];
+                auto tile = openloco::map::map_pos3({ town->x, town->y, tileZ });
+                auto zoomLevel = openloco::ui::viewportmgr::ZoomLevel::half;
+                viewportmgr::create(self, 0, gfx::point_t(widget->top + self->y, widget->left + self->x), gfx::ui_size_t(widget->width(), widget->height()), zoomLevel, tile);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: I think the size needs to be -{2, 2}, and the offset needs to be +{1,1}, to take in account the border.

File: src/openloco/windows/townwnd.cpp [hunk 2]
```diff
@@ -1,31 +1,791 @@
...
+                string_ids::town_size_town,
+                string_ids::town_size_city,
+                string_ids::town_size_metropolis,
+            };
+
+            auto town = townmgr::get(self->number);
+            commonFormatArgs[0] = townSizeLabelIds[(uint8_t)town->size];
+            *(int32_t*)&commonFormatArgs[1] = town->population;
+
+            const uint16_t xPos = self->x + self->widgets[widx::unk_8].left;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
            const uint16_t xPos = self->x + self->widgets[widx::unk_8].left - 1;
```

File: src/openloco/windows/townwnd.cpp [hunk 3]
```diff
@@ -1,31 +1,791 @@
...
+            const uint16_t yPos = self->y + self->widgets[widx::unk_8].top;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
            const uint16_t yPos = self->y + self->widgets[widx::unk_8].top - 1;
```

File: src/openloco/windows/townwnd.cpp [hunk 4]
```diff
@@ -1,31 +1,791 @@
...
+            const uint16_t width = self->widgets[widx::unk_8].right - self->widgets[widx::unk_8].left;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
            const uint16_t width = self->widgets[widx::unk_8].width() - 1;
```

File: src/openloco/windows/townwnd.cpp [hunk 5]
```diff
@@ -1,31 +1,791 @@
...
+                case widx::centre_on_viewport:
+                    call(0x0049932D, regs);
+                    break;
+
+                case widx::expand_town:
+                    call(0x004990B9, regs);
+                    break;
+
+                case widx::demolish_town:
+                    call(0x0049916A, regs);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Only call. Implement.

File: src/openloco/windows/townwnd.cpp [hunk 6]
```diff
@@ -1,31 +1,791 @@
...
+                case common::widx::tab_company_ratings:
+                    common::switchTab(self, widgetIndex);
+                    break;
+
+                case widx::centre_on_viewport:
+                    call(0x0049932D, regs);
+                    break;
+
+                case widx::expand_town:
+                    call(0x004990B9, regs);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Only call. Implement.

File: src/openloco/windows/townwnd.cpp [hunk 7]
```diff
@@ -1,31 +1,791 @@
...
+                    break;
+
+                case common::widx::tab_town:
+                case common::widx::tab_population:
+                case common::widx::tab_company_ratings:
+                    common::switchTab(self, widgetIndex);
+                    break;
+
+                case widx::centre_on_viewport:
+                    call(0x0049932D, regs);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Only call. Implement.

File: src/openloco/windows/townwnd.cpp [hunk 8]
```diff
@@ -1,31 +1,791 @@
...
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+
+            gfx::drawpixelinfo_t* clipped = nullptr;
+            if (!gfx::clip_drawpixelinfo(&clipped, dpi, self->x, self->y + 44, self->width, self->height - 44))
+                return;
+
+            auto town = townmgr::get(self->number);
+            *(int32_t*)&*commonFormatArgs = town->history_min_population;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Would rather have this in its own variable, and set the formatarg for each label

File: src/openloco/windows/townwnd.cpp [hunk 9]
```diff
@@ -1,31 +1,791 @@
...
+
+            // Draw Y label and grid lines.
+            for (int16_t yPos = self->height - 57; yPos > 14; yPos -= 20)
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
            for (int16_t yPos = self->height - 57; yPos >= 14; yPos -= 20)
```

File: src/openloco/windows/townwnd.cpp [hunk 10]
```diff
@@ -1,31 +1,791 @@
...
+            {
+                const uint16_t xPos = 39;
+                gfx::fill_rect(clipped, xPos, yPos, xPos + 241, yPos, colour::get_shade(self->colours[1], 4));
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Use `draw_rect` signature instead

File: src/openloco/windows/townwnd.cpp [hunk 11]
```diff
@@ -1,31 +1,791 @@
...
+            if (!gfx::clip_drawpixelinfo(&clipped, dpi, self->x, self->y + 44, self->width, self->height - 44))
+                return;
+
+            auto town = townmgr::get(self->number);
+            *(int32_t*)&*commonFormatArgs = town->history_min_population;
+
+            // Draw Y label and grid lines.
+            for (int16_t yPos = self->height - 57; yPos > 14; yPos -= 20)
+            {
+                const uint16_t xPos = 39;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Base value is 41. 39 is the value used for the label.

File: src/openloco/date.h [hunk 12]
```diff
@@ -39,6 +39,7 @@ namespace openloco
     };
 
     uint32_t current_day();
+    uint8_t current_month();
```
Comment (marijnvdwerf) [inline src/openloco/date.h]: Use `month_id` enum

File: src/openloco/game_commands.h [hunk 13]
```diff
@@ -26,6 +26,18 @@ namespace openloco::game_commands
         do_command(21, regs);
     }
 
+    inline void do_46(uint8_t bl, uint16_t cx, uint16_t ax, uint32_t edx, uint32_t ebp, uint32_t edi)
```
Comment (marijnvdwerf) [inline src/openloco/game_commands.h]: Would rather see this as a single game command with memcpys instead of pointer magic.

e.g. `do_46(town_id cx, char[36])`

File: src/openloco/graphics/gfx.cpp [hunk 14]
```diff
@@ -747,6 +772,19 @@ namespace openloco::gfx
         fill_rect_inset(dpi, x, y, x + dx - 1, y + dy - 1, colour, flags);
     }
 
+    // 0x00452DA4
+    void draw_line(gfx::drawpixelinfo_t* dpi, int16_t left, int16_t top, int16_t right, int16_t bottom, uint32_t color)
```
Comment (marijnvdwerf) [inline src/openloco/graphics/gfx.cpp]:
```suggestion
    void draw_line(gfx::drawpixelinfo_t* dpi, point_t start, point_t stop, uint8_t color)
```

Comment (marijnvdwerf) [reply src/openloco/graphics/gfx.cpp]: Also: colour.

File: src/openloco/windows/townwnd.cpp [hunk 15]
```diff
@@ -1,31 +1,791 @@
...
+
+                int16_t rating = (std::clamp<int16_t>(town->company_ratings[i], -1000, 1000) + 1000) / 20;
+                string_id rank{};
+                if (rating >= 70)
+                    rank = string_ids::town_rating_excellent;
+                else if (rating >= 60)
+                    rank = string_ids::town_rating_good;
+                else if (rating >= 50)
+                    rank = string_ids::town_rating_average;
+                else if (rating >= 24)
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
                else if (rating >= 25)
```

File: src/openloco/windows/townwnd.cpp [hunk 16]
```diff
@@ -1,31 +1,791 @@
...
+                    rank = string_ids::town_rating_poor;
+                else
+                    rank = string_ids::town_rating_appalling;
+
+                commonFormatArgs[0] = companymgr::get(i)->name;
+                commonFormatArgs[2] = rating;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Any idea why this skips argument 1?

File: src/openloco/windows/townwnd.cpp [hunk 17]
```diff
@@ -1,31 +1,791 @@
...
+            }
+        }
+
+        // 0x00499287
+        static void text_input(window* self, widget_index callingWidget, char* input)
+        {
+            if (callingWidget != common::widx::caption)
+                return;
+
+            if (*input == '\0')
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: strlen() == 0

File: src/openloco/windows/townwnd.cpp [hunk 18]
```diff
@@ -1,31 +1,791 @@
...
+
+                    gfx::fill_rect(clipped, xPos, 11, xPos, self->height - 57, colour::get_shade(self->colours[1], 4));
+                }
+
+                // Draw population graph
+                uint16_t yPos1 = -town->history[i] + (self->height - 57);
+                uint16_t yPos2 = -town->history[i + 1] + (self->height - 57);
+                gfx::draw_line(clipped, xPos, yPos1, xPos + 1, yPos2, colour::get_shade(self->colours[1], 7));
+
+                if (--month < 0)
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Not a fan of having `(--var == value)`-type checks. These operators are short, but really difficult to read (https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md#disadvantages-of-these-operators)

File: src/openloco/windows/townwnd.cpp [hunk 19]
```diff
@@ -1,31 +1,791 @@
...
+
+            for (uint8_t i = town->history_size - 1; i > 0; i--)
+            {
+                const uint16_t xPos = 41 + i;
+                const uint16_t yPos = 2;
+
+                // Draw horizontal year and vertical grid lines.
+                if (month == 0)
+                {
+                    if (yearSkip == 0)
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: `if ((year %3) == (current_year() % 3)` (with the latter part stored in a variable) would remove the need for a `yearSkip` variable.

File: src/openloco/windows/townwnd.cpp (hunk 11 above)
Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: Yes, this was a deliberate deviation on my part — I've adjusted the calls below accordingly. Saves the offset with two on the call below.

File: src/openloco/windows/townwnd.cpp (hunk 19 above)
Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: That requires doing two divisions though, so I understand why Sawyer did it as-is…

File: src/openloco/windows/townwnd.cpp (hunk 16 above)
Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: No, I thought it strange, too, especially since `town_rating_company_percentage_rank` reads:

```
1839: "{COLOUR WINDOW_2}{STRINGID}: {COLOUR BLACK}{INT16}% ({STRINGID})"
```

So you'd expect it to not skip anything. But it does, as you've no doubt seen in IDA.

File: src/openloco/graphics/gfx.cpp (hunk 14 above)
Comment (AaronVanGeffen) [reply src/openloco/graphics/gfx.cpp]: `fill_rect`, `fill_rect_inset` and `draw_rect` have committed similar transgressions, heh. Will correct.

File: src/openloco/game_commands.h (hunk 13 above)
Comment (AaronVanGeffen) [reply src/openloco/game_commands.h]: I would prefer to see that, too, but preferably not in this PR.

Comment (AaronVanGeffen) [reply src/openloco/game_commands.h]: I would prefer to see that, too, but preferably not in this PR. This will be difficult one to tackle, as these commands are (ultimately) networked. The same char pointer will not be valid for other players.

File: src/openloco/windows/townwnd.cpp (hunk 10 above)
Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: The `draw_rect` signature assumes `dx` and `dy` to be positive. If you want to draw a horizontal or vertical line, this will go wrong:
```
draw_rect_impl(dpi, x, y, x + dx - 1, y + dy - 1, colour);
```
I can work around this by using 1 for `dx` and `dy`, but that seems wrong.

Comment (marijnvdwerf) [reply src/openloco/windows/townwnd.cpp]: Why? A rectangle with height = 1 is a line.

File: src/openloco/widget.cpp [hunk 20]
```diff
@@ -51,6 +51,15 @@ namespace openloco::ui::widget
         call(0x004CF3EB, regs);
     }
 
+    void sub_4CF487(gfx::drawpixelinfo_t* dpi, const window* window, const widget_index index)
```
Comment (marijnvdwerf) [inline src/openloco/widget.cpp]: This draws the viewport center button

Comment (AaronVanGeffen) [reply src/openloco/widget.cpp]: Ah, I thought I'd already renamed this one. Indeed.

File: src/openloco/ui/WindowManager.h [hunk 21]
```diff
@@ -191,6 +189,11 @@ namespace openloco::ui::tooltip
     void update(ui::window* window, int32_t widgetIndex, string_id stringId, int16_t x, int16_t y);
 }
 
+namespace openloco::ui::windows::town
+{
+    window* open(uint16_t townId);
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openloco/ui/WindowManager.h]: `town_id_t` is defined in `town.h`. We should probably consider moving these types to a more common location so we don't have to include the entire header (and its dependencies) for one little type.

File: src/openloco/graphics/gfx.cpp [hunk 22]
```diff
@@ -705,37 +730,38 @@ namespace openloco::gfx
...
     {
         registers regs;
         regs.ax = left;
         regs.bx = right;
         regs.cx = top;
         regs.dx = bottom;
-        regs.ebp = color;
+        regs.ebp = colour;
         regs.edi = (uint32_t)dpi;
+        regs.si = flags;
```
Comment (marijnvdwerf) [inline src/openloco/graphics/gfx.cpp]: I doubt this is correct (`si` gets overwritten at `0x4474D7`). Why did you add this?

File: src/openloco/town.h [hunk 23]
```diff
@@ -32,11 +32,13 @@ namespace openloco
     struct town
     {
         string_id name;
-        uint8_t pad_02[0x06 - 0x02];
+        int16_t x;
```
Comment (marijnvdwerf) [inline src/openloco/town.h]: `tile_coord_t`

File: src/openloco/windows/townwnd.cpp [hunk 24]
```diff
@@ -1,31 +1,863 @@
...
+                    common::switchTab(self, widgetIndex);
+                    break;
+
+                case widx::centre_on_viewport:
+                {
+                    if (self->viewports[0] == nullptr || self->saved_view.isEmpty())
+                        break;
+
+                    registers regs;
+                    if (self->saved_view.hasUnkFlag())
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Will never happen

Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: You'll have to elaborate here. There's an explicit check for this at `0x0049934C`, and normally you're rather strict about these things…

File: src/openloco/graphics/gfx.cpp (hunk 22 above)
Comment (AaronVanGeffen) [reply src/openloco/graphics/gfx.cpp]: `0x004CF581` and `0x004CF58F` are setting the register, I'm just passing it on.

File: src/openloco/game_commands.h [hunk 25]
```diff
@@ -39,6 +51,15 @@ namespace openloco::game_commands
         do_command(71, regs);
     }
 
+    inline bool do_50(uint8_t bl, uint8_t townId)
```
Comment (marijnvdwerf) [inline src/openloco/game_commands.h]:
```suggestion
    inline bool do_50(uint8_t bl, town_id_t townId)
```

File: src/openloco/windows/townwnd.cpp [hunk 26]
```diff
@@ -1,31 +1,863 @@
...
+            // Call to sub_498E9B has been deliberately omitted.
+
+            self->set_size(gfx::ui_size_t(192, 161), gfx::ui_size_t(600, 440));
+
+            common::on_resize(self);
+
+            if (self->viewports[0] != nullptr)
+            {
+                uint16_t newWidth = self->width - 30;
+                if (is_editor_mode())
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
                if (!is_editor_mode())
```

File: src/openloco/game_commands.h (hunk 25 above)
Comment (AaronVanGeffen) [reply src/openloco/game_commands.h]: I agree, but as I said in another comment:

> `town_id_t` is defined in `town.h`. We should probably consider moving these types to a more common location so we don't have to include the entire header (and its dependencies) for one little type.
> 
> … Preferably in a different PR.

I think we should move a bunch such types to e.g. `types.hpp` for easy inclusion.

File: src/openloco/graphics/gfx.cpp (hunk 22 above)
Comment (marijnvdwerf) [reply src/openloco/graphics/gfx.cpp]: They call the `_inset` version though.

Comment (AaronVanGeffen) [reply src/openloco/graphics/gfx.cpp]: I see what you mean now. You're right, they did not need to be added.

File: src/openloco/widget.cpp [hunk 27]
```diff
@@ -51,6 +51,24 @@ namespace openloco::ui::widget
         call(0x004CF3EB, regs);
     }
 
+    // 0x004CF487
+    void drawViewportCentreButton(gfx::drawpixelinfo_t* dpi, const window* window, const widget_index widgetIndex)
+    {
+        auto& widget = window->widgets[widgetIndex];
+        if (input::is_hovering(window->type, window->number, widgetIndex))
+        {
+            gfx::fill_rect(dpi, widget.left + window->x, widget.top + window->y, widget.right + window->x, widget.bottom + window->y, gfx::recolour(0x34));
```
[the remark this answers is not in the document]
Comment (marijnvdwerf) [reply src/openloco/widget.cpp]: Also, the second rectangle drawn is missing

File: src/openloco/widget.cpp [hunk 28]
```diff
@@ -51,6 +51,24 @@ namespace openloco::ui::widget
...
+
+            uint8_t flags = 0;
+            if (input::state() == input::input_state::widget_pressed && input::is_pressed(window->type, window->number, widgetIndex))
+                flags = 0x20;
+
+            gfx::fill_rect_inset(dpi, widget.left + window->x, widget.top + window->y, widget.right + window->x, widget.bottom + window->y, window->colours[1] | 0x80, flags);
```
Comment (marijnvdwerf) [inline src/openloco/widget.cpp]:
```suggestion
            gfx::draw_rect_inset(dpi, widget.left + window->x, widget.top + window->y, widget.width(), widget.height(), colour::translucent(window->colours[1]), flags);
```

File: src/openloco/widget.cpp [hunk 29]
```diff
@@ -51,6 +51,24 @@ namespace openloco::ui::widget
...
+    // 0x004CF487
+    void drawViewportCentreButton(gfx::drawpixelinfo_t* dpi, const window* window, const widget_index widgetIndex)
+    {
+        auto& widget = window->widgets[widgetIndex];
+        if (input::is_hovering(window->type, window->number, widgetIndex))
+        {
+            gfx::fill_rect(dpi, widget.left + window->x, widget.top + window->y, widget.right + window->x, widget.bottom + window->y, gfx::recolour(0x34));
+
+            uint8_t flags = 0;
+            if (input::state() == input::input_state::widget_pressed && input::is_pressed(window->type, window->number, widgetIndex))
```
Comment (marijnvdwerf) [inline src/openloco/widget.cpp]: `input::state() == input::input_state::widget_pressed` already gets handled by `input::is_pressed`

File: src/openloco/widget.cpp [hunk 30]
```diff
@@ -51,6 +51,24 @@ namespace openloco::ui::widget
...
+                flags = 0x20;
+
+            gfx::fill_rect_inset(dpi, widget.left + window->x, widget.top + window->y, widget.right + window->x, widget.bottom + window->y, window->colours[1] | 0x80, flags);
+        }
+
+        gfx::draw_image(dpi, widget.left + window->x, widget.top + window->y, gfx::recolour(image_ids::centre_viewport));
```
Comment (marijnvdwerf) [inline src/openloco/widget.cpp]:
```suggestion
        gfx::draw_image(dpi, widget.left + window->x, widget.top + window->y, gfx::recolour(image_ids::centre_viewport, window->colours[1]));
```

File: src/openloco/window.cpp [hunk 31]
```diff
@@ -105,6 +105,16 @@ namespace openloco::ui
         return (this->holdable_widgets & (1ULL << index)) != 0;
     }
 
+    // 0x0045A0B3
+    void window::drawViewports(gfx::drawpixelinfo_t* dpi)
+    {
+        if (viewports[0] != nullptr)
+            viewports[0]->render(dpi, gfx::point_t(dpi->x, dpi->y), gfx::ui_size_t(dpi->width, dpi->height));
```
Comment (marijnvdwerf) [inline src/openloco/window.cpp]: I don't really see a point for passing the `gfx::point_t`s here, as they are properties of the `dpi` anyway.

File: src/openloco/window.cpp [hunk 32]
```diff
@@ -1062,13 +1070,9 @@ namespace openloco::ui
...
         {
-            if (this->type == addr<0x0052336F, WindowType>() && this->number == addr<0x00523370, uint16_t>())
-            {
-                if (input::has_flag((input::input_flags)(1 << 0)))
-                {
-                    pressed_widget = 1ULL << addr<0x00523372, uint32_t>();
-                }
-            }
+            widget_index widgetIndex = input::get_pressed_widget_index();
+            if (input::is_pressed(type, number, widgetIndex))
```
Comment (marijnvdwerf) [inline src/openloco/window.cpp]: Adding an `bool is_pressed(ui::WindowType type, ui::window_number number)` function would make more sense to me.

File: src/openloco/window.h [hunk 33]
```diff
@@ -408,16 +325,22 @@ namespace openloco::ui
         int16_t var_88C;
 
         window(gfx::point_t position, gfx::ui_size_t size);
-        constexpr void set_size(gfx::ui_size_t size)
+
+        constexpr void set_size(gfx::ui_size_t minSize, gfx::ui_size_t maxSize)
```
Comment (marijnvdwerf) [inline src/openloco/window.h]: I'd be fine with this method, if it also did the clamping of the current size. If you don't want to do that, the method name needs to be changed.

File: src/openloco/windows/townwnd.cpp (hunk 24 above)
Comment (marijnvdwerf) [reply src/openloco/windows/townwnd.cpp]: The original code has dead branches, since a town window will never focus on a `thing`. `SavedView` is the same as `edx` and `ecx` in `viewportmgr::create(registers regs, int index)`. It's basically used as a hash, in case the postion of the town (?) changes.

File: src/openloco/windows/townwnd.cpp [hunk 34]
```diff
@@ -1,31 +1,863 @@
...
+                        regs.dx = self->saved_view.surfaceZ;
+                    }
+
+                    regs.esi = (int32_t)WindowManager::getMainWindow();
+                    call(0x004C6827, regs);
+                    break;
+                }
+
+                case widx::expand_town:
+                {
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Add a comment to the bigger ones with their address (`0x004990B9`)

File: src/openloco/windows/townwnd.cpp [hunk 35]
```diff
@@ -1,31 +1,863 @@
...
+                        }
+
+                        tempYear += 7;
+                        set_current_year(tempYear);
+                    }
+
+                    set_current_year(currentYear);
+
+                    // Set new history entry.
+                    uint8_t historyEntry = std::min<uint8_t>(town->population / 50, 255);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: Just to be sure, does `town->population / 50` get clamped to an `uint8_t` before or after clamping of the value? I understand that it does't matter in this case, since it's unsigned. But would still like to know.

File: src/openloco/windows/townwnd.cpp [hunk 36]
```diff
@@ -1,31 +1,863 @@
...
-            w->enabled_widgets &= ~(1 << 1);
+            // Call to sub_498E9B has been deliberately omitted.
+
+            self->set_size(gfx::ui_size_t(192, 161), gfx::ui_size_t(600, 440));
+
+            common::on_resize(self);
+
+            if (self->viewports[0] != nullptr)
+            {
+                uint16_t newWidth = self->width - 30;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: can't we use `self->widgets[widx::viewport].width()` for these?

File: src/openloco/windows/townwnd.cpp [hunk 37]
```diff
@@ -1,31 +1,863 @@
...
+                if (!is_editor_mode())
+                    newWidth += 22;
+
+                uint16_t newHeight = self->height - 59;
+
+                auto& viewport = self->viewports[0];
+                if (newWidth != viewport->width || newHeight != viewport->height)
+                {
+                    viewport->width = newWidth;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: This will all get reset when `initViewport` gets called after doing `self->saved_view.clear();`

File: src/openloco/windows/townwnd.cpp [hunk 38]
```diff
@@ -1,31 +1,863 @@
...
+                auto widget = &self->widgets[widx::viewport];
+                auto tile = openloco::map::map_pos3({ town->x, town->y, tileZ });
+                auto zoomLevel = openloco::ui::viewportmgr::ZoomLevel::half;
+                viewportmgr::create(self, 0, gfx::point_t(widget->left + self->x + 1, widget->top + self->y + 1), gfx::ui_size_t(widget->width() - 2, widget->height() - 2), zoomLevel, tile);
+                self->invalidate();
+                self->flags |= window_flags::flag_2;
+            }
+            // 0x00499B39 end
+
+            if (self->viewports[0]->x != 0)
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]:
```suggestion
            if (self->viewports[0] != nullptr)
```

File: src/openloco/windows/townwnd.cpp [hunk 39]
```diff
@@ -1,31 +1,863 @@
...
+
+            // 0x00499B39 start
+            if (self->viewports[0] == nullptr)
+            {
+                auto widget = &self->widgets[widx::viewport];
+                auto tile = openloco::map::map_pos3({ town->x, town->y, tileZ });
+                auto zoomLevel = openloco::ui::viewportmgr::ZoomLevel::half;
+                viewportmgr::create(self, 0, gfx::point_t(widget->left + self->x + 1, widget->top + self->y + 1), gfx::ui_size_t(widget->width() - 2, widget->height() - 2), zoomLevel, tile);
+                self->invalidate();
+                self->flags |= window_flags::flag_2;
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: `flag_2` is equal to `WF_NO_SCROLLING` in OpenRCT2.

File: src/openloco/windows/townwnd.cpp [hunk 40]
```diff
@@ -1,31 +1,863 @@
...
+            if (input::is_tool_active(window->type, window->number))
+                input::cancel_tool();
+
+            window = WindowManager::bringToFront(WindowType::town, townId);
+        }
+
+        if (window == nullptr)
+        {
+            // 0x00499C0D start
+            window = WindowManager::createWindow(WindowType::town, windowSize, 0, &town::events);
```
Comment (marijnvdwerf) [inline src/openloco/windows/townwnd.cpp]: `1 << 8` is passed as a flag.

Comment (marijnvdwerf) [reply src/openloco/windows/townwnd.cpp]: Remove all properties that are set later on, or have a default value.

File: src/openloco/windows/townwnd.cpp (hunk 35 above)
Comment (AaronVanGeffen) [reply src/openloco/windows/townwnd.cpp]: Both arguments are interpreted as a value (object) of the type (class) passed. As they are rvalues, this re-interpretation takes place after division. Consider:
```
template<class T> 
const T& min(const T& a, const T& b)
{
    return (b < a) ? b : a;
}
```
(via [cppref](https://en.cppreference.com/w/cpp/algorithm/min#Possible_implementation))

File: src/openloco/window.cpp (hunk 32 above)
Comment (AaronVanGeffen) [reply src/openloco/window.cpp]: I don't disagree, and will change it, but we'll still need the widget index anyway to set `pressed_widget`.

**Comment by AaronVanGeffen** [comment]
Right, I think that's all comments taken care off. Please tell me it's good to merge now…

File: src/openloco/window.h [hunk 41]
```diff
@@ -174,12 +175,13 @@ namespace openloco::ui
     {
         constexpr uint32_t stick_to_back = 1 << 0;
         constexpr uint32_t stick_to_front = 1 << 1;
-        constexpr uint32_t flag_2 = 1 << 2;
+        constexpr uint32_t no_scrolling = 1 << 2;
```
Comment (marijnvdwerf) [inline src/openloco/window.h]: Add some clarification that this is about viewport scrolling
